### DISCUSSION TITLE
Connessione remota Home Assistant NAT

### DISCUSSION THREAD

**Discussion by chopin94** [indomus/forum#2494] [Networking]
Salve a tutti, vi spiego il mio problema nella speranza di venirne a capo.

Per entrare remotamente sul mio HA ho deciso di utilizzare il metodo in cui tramite una vm su GCP (google) con openVPN accedo anche se sono sotto NAT.

Non so cosa sia successo o se già era così precedentemente ma noto che

- con l'indirizzo http://dominio.duckdns.org:8123 riesco ad entrare da fuori
- appena attivo caddy e metto i certificati finche nel caddyfile uso "Saluti da indomus" lui funziona nel momento in cui imposto il reverse proxy al mio container :8123 non funziona 
- in più la cosa realmente assurda è che se io dal browser del mio server accedo a http://192.168.1.x:9000 per portainer non funziona, mentre con 127.0.0.1:9000 sì. 

Sono bloccato avete consigli o spunti per proseguire?

**Comment by snakuzzo** [comment]
Da come hai descritto il problema, immagino tu non sia abbastanza pratico, per cui ti consiglierei vivamente di seguire la procedura utilizzando cloudflared.

Detto ciò, le cose non è che cambiano da sole.

Per quanto riguarda Caddy ed HA, se non hai problemi con il test per l'accesso in https con caddy "saluti da indomus", vuol dir che sei a metà dell'opera. Quanto ad HA, il problema è che non puoi dire semplicemente "non funziona"...devi dettagliare meglio

Per Portiner invece, vuol dire che sullo stesso server puoi accedere solo utilizzando la loopback. Qui stesso discorso di cui sopra...devi dettagliare il "non funziona"

**Comment by chopin94** [comment]
Ciao diego ho preferito scrivere qua in modo da essere più chiaro considerando la mia incapacità. 


per quanto riguarda portainer, vale anche per frigate ma non per esphome e z2m.... mah


Quale altre informazioni posso dare per chiarire?

**Reply by snakuzzo** [reply]
Quando fai reply, fallo sotto l'ultimo messaggio. Non come hai fatto ora che hai creato un nuovo thread 

Portainer e frigate li raggiungi da altri ip sulla stessa LAN? 

Poi queste cose con il problema che hai con caddy ed HA non c'entrano assolutamente niente.
Quando provi ad accedere da internet sulla porta dedicata ad HA qual è il comportamento?
Che vedi dal browser del client da cui ti connetti?

**Reply by snakuzzo** [reply]
Quella guida comunque è troppo articolata per chi non sa bene cosa sta facendo...e pure per noi diventa complicato dare supporto in questo modo...il consiglio sul lasciar perdere in favore di Cloudflared resta sempre valido

**Reply by chopin94** [reply]
Ricevo http error 502
e sul log di home assistant:

2023-06-08 18:42:13.107 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
File "/usr/local/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 332, in data_received
messages, upgraded, tail = self._request_parser.feed_data(data)
File "aiohttp/_http_parser.pyx", line 551, in aiohttp._http_parser.HttpParser.feed_data
aiohttp.http_exceptions.BadStatusLine: 400, message="Bad status line 'Invalid method encountered'"

**Reply by chopin94** [reply]
> Quella guida comunque è troppo articolata per chi non sa bene cosa sta facendo...e pure per noi diventa complicato dare supporto in questo modo...il consiglio sul lasciar perdere in favore di Cloudflared resta sempre valido

Si lo capisco dieee però mi rode che credo di essere molto vicino al problema. Vorrei provarci un pò prima di arrendermi.

**Reply by chopin94** [reply]
log di caddy:

ERR ts=1686244502.411313 logger=http.log.error msg=dial tcp 192.168.1.100:8123: i/o timeout request={"remote_ip":"10.154.244.1","remote_port":"53295","proto":"HTTP/2.0","method":"GET","host":"xxx.duckdns.org","uri":"/","headers":{"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7"],"Sec-Fetch-Site":["none"],"Sec-Fetch-Dest":["document"],"Sec-Ch-Ua-Mobile":["?0"],"User-Agent":["Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36 Edg/114.0.1823.37"],"Sec-Ch-Ua-Platform":["\"Windows\""],"Upgrade-Insecure-Requests":["1"],"Sec-Fetch-Mode":["navigate"],"Sec-Fetch-User":["?1"],"Accept-Encoding":["gzip, deflate, br"],"Accept-Language":["it"],"Cache-Control":["max-age=0"],"Sec-Ch-Ua":["\"Not.A/Brand\";v=\"8\", \"Chromium\";v=\"114\", \"Microsoft Edge\";v=\"114\""]},"tls":{"resumed":true,"version":772,"cipher_suite":4867,"proto":"h2","server_name":"xxx.duckdns.org"}} duration=3.001836653 status=502 err_id=ded5iax2f err_trace=reverseproxy.statusError (reverseproxy.go:1272)

**Reply by snakuzzo** [reply]
Quell'errore HTTP 502 non è bello.
Il giro è questo...


- La VM GCP che ha IP pubblico è raggiungibile tramite FQDN duckdns (script in cron per aggiornare il record A)
- Una volta arrivati sulla VM GCP c'è una regola che inoltra i pacchetti TCP sulla porta da te utilizzata (9000 ???) verso l'IP assegnato dalla VPN della VM locale (10.x.x.x) sulla porta con cui Caddy espone il servizio proxato di HA (9000???)
- Caddy inoltra il pacchetto verso l'IP LAN di Home assistant (perché immagino che il container sia stato creato con network in  modalità bridge esponendo le singole porte) 

```
INTERNET -> FQDN duckdns (9000) -> VM GCP -> regola iptables (9000->9000 ip 10.x.x.x) ->  tunnel OpenVPN -> VM locale -> Caddy (9000) -> Home Assistant (http://192.168.1.100:8123)
```

Da qui dovresti capire cos'è che hai sbagliato a fare.
Nel caso condividi:

- regole iptables VM GCP
- Configurazione Caddy (container e Caddyfile)
- Configurazione HA (container e configuration.yaml)

Tutto quello che è codice va scritto usando il markdown. Se non lo utilizzi (tipo come hai fatto nei precedenti messaggi) cancello il post.

**Reply by chopin94** [reply]
Ho risoltoooooooooo

Praticamente nel caddyfile dove scrivo reverse proxy non dovevo inserire 192.168.x.x, ma la subnet della vpn e adesso funziona perfettamente.
Putroppo non riesco a parlare un linguaggio tecnico adeguato per farmi comprendere per cui capisco che sia difficile capirmi.

**Reply by snakuzzo** [reply]
Scusa eh...ma caddy dove l'hai installato ? Sulla VM GCP o in locale ?
Perché questa cosa che hai fatto ha senso solo se hai messo Caddy sulla VM GCP
Se ce l'hai in locale, non ha alcun senso quello che hai fatto...e se funziona solo così, vuol dire che hai altri problemi

**Reply by chopin94** [reply]
Allora ho altri problemi. Ma non so proprio da dove cominciare. Intanto funziona, così posso ragionarci con calma. Praticamente ormai se non ho accesso da fuori a casa sbiello. 

Potrebbe anche dipendere da qualcosa che avevo toccato nelle route statiche qualche settimana fa.... 

Troppe variabili dieee

**Reply by snakuzzo** [reply]
io non posso dirtelo...posso solo dirti che hai sbagliato. Ti sta funzionando per "botta di culo" :)
Ci sta che non vuoi arrenderti, ma non ci sta che risolvi i problemi in modo sbagliato. Quindi ti rinnovo l'invito ad usare Cloudflared, che ti dà altri innumerevoli vantaggi

**Reply by chopin94** [reply]
Comunque sono partito da un'installazione pulita rifatto il tutto e funziona senza alcun problema. C'era sicuramente problema di rete da parte del server. 

"in più la cosa realmente assurda è che se io dal browser del mio server accedo a http://192.168.1.x:9000/ per portainer non funziona, mentre con 127.0.0.1:9000 sì."

Questo già mi faceva sorgere il sospetto aahhahahah
